tests/node/test_wakunode_sharding.nim

```diff
@@ -0,0 +1,1026 @@
+      let
+        serverResult2 = await serverHandler.waitForResult(FUTURE_TIMEOUT)
+        clientResult2 = await clientHandler.waitForResult(FUTURE_TIMEOUT)
+
+      # Then the client receives the message but the server does not
+      check serverResult2.isErr()
+      assertResultOk(clientResult2)
+
+  suite "Automatic Sharding Mechanics":
+    asyncTest "Content Topic-Based Shard Dialing":
+      # Given a connected server and client subscribed to the same content topic (with two different formats)
+      let
+        contentTopicShort = "/toychat/2/huilong/proto"
+        contentTopicFull = "/0/toychat/2/huilong/proto"
+        pubsubTopic = "/waku/2/rs/0/58355"
+          # Automatically generated from the contentTopic above
+
+      let
+        serverHandler = server.subscribeToContentTopicWithHandler(contentTopicShort)
+        clientHandler = client.subscribeToContentTopicWithHandler(contentTopicFull)
+
+      await client.connectToNodes(@[server.switch.peerInfo.toRemotePeerInfo()])
+
+      # When the client publishes a message
+      discard await client.publish(
+        some(pubsubTopic),
+        WakuMessage(payload: "message1".toBytes(), contentTopic: contentTopicShort),
+      )
+      let
+        serverResult1 = await serverHandler.waitForResult(FUTURE_TIMEOUT_LONG)
+        clientResult1 = await clientHandler.waitForResult(FUTURE_TIMEOUT_LONG)
+
+      # Then the server and client receive the message
+      assertResultOk(serverResult1)
+      assertResultOk(clientResult1)
+
+      # When the server publishes a message
+      serverHandler.reset()
+      clientHandler.reset()
+      discard await server.publish(
+        some(pubsubTopic),
+        WakuMessage(payload: "message2".toBytes(), contentTopic: contentTopicFull),
+      )
+      let
+        serverResult2 = await serverHandler.waitForResult(FUTURE_TIMEOUT)
+        clientResult2 = await clientHandler.waitForResult(FUTURE_TIMEOUT)
+
+      # Then the client and server receive the message
+      assertResultOk(serverResult2)
+      assertResultOk(clientResult2)
+
+    asyncTest "Exclusion of Irrelevant Autosharded Topics":
+      # Given a connected server and client subscribed to different content topics
+      let
+        contentTopic1 = "/toychat/2/huilong/proto"
+        pubsubTopic1 = "/waku/2/rs/0/58355"
+          # Automatically generated from the contentTopic above
+        contentTopic2 = "/0/toychat2/2/huilong/proto"
+        pubsubTopic2 = "/waku/2/rs/0/23286"
+          # Automatically generated from the contentTopic above
+
+      let
+        serverHandler = server.subscribeToContentTopicWithHandler(contentTopic1)
+        clientHandler = client.subscribeToContentTopicWithHandler(contentTopic2)
+
+      await client.connectToNodes(@[server.switch.peerInfo.toRemotePeerInfo()])
+
+      # When the server publishes a message in the server's subscribed topic
+      discard await server.publish(
+        some(pubsubTopic1),
```

Comment:
You did a great job of calculating the value and then hard-coding it above but I wonder if we could use a value obtained from codebase logic instead of using a hard-coded value. That would also help to understand how the pubsub topic is inferred from the content topic